File: .dev/sniff-coverage-matrix.php
#!/usr/bin/env php
<?php declare(strict_types = 1);

$root = dirname(__DIR__);
$reportsDir = $root . '/.reports';
$sniffsRoot = $root . '/tests/Sniffs';

$phpcsOutput = [];
$exitCode = 0;
exec('vendor/bin/phpcs --standard=ruleset.xml -e 2>&1', $phpcsOutput, $exitCode);

if ($exitCode !== 0) {
	fwrite(STDERR, "Failed to list enabled sniffs.\n");
	fwrite(STDERR, implode("\n", $phpcsOutput) . "\n");
	exit($exitCode);
}

$enabledSniffs = [];
foreach ($phpcsOutput as $line) {
	if (preg_match('#^\s*([A-Z][A-Za-z0-9]*\.[A-Za-z0-9_]+\.[A-Za-z0-9_]+)\s*$#', $line, $matches) === 1) {
		$enabledSniffs[$matches[1]] = true;
	}
}
$enabledSniffs = array_keys($enabledSniffs);
sort($enabledSniffs);

$refDirectories = [];
$iterator = new RecursiveIteratorIterator(new RecursiveDirectoryIterator($sniffsRoot));
foreach ($iterator as $file) {
	if (!$file->isFile() || !str_ends_with($file->getFilename(), '.ruleset.xml')) {
		continue;
	}

	$content = file_get_contents($file->getPathname());
	if ($content === false) {
		continue;
	}

	if (preg_match_all('#<rule\s+ref="([^"]+)"#', $content, $matches) !== false) {
		foreach ($matches[1] as $ref) {
			if (!str_contains($ref, '.')) {
				continue;
			}

			$refDirectories[$ref] ??= [];
			$refDirectories[$ref][$file->getPath()] = true;
		}
	}
}

foreach ($refDirectories as &$directories) {
	$directories = array_keys($directories);
	sort($directories);
}
unset($directories);

if (!is_dir($reportsDir)) {
	mkdir($reportsDir, 0777, true);
}

$csvPath = $reportsDir . '/sniff-coverage-matrix.csv';
$summaryPath = $reportsDir . '/sniff-coverage-summary.md';

$rows = [];
foreach ($enabledSniffs as $sniff) {
	$directories = $refDirectories[$sniff] ?? [];
	$fixtures = [];
	$hasGood = false;
	$hasBad = false;
	$customCount = 0;

	foreach ($directories as $directory) {
		$phpFiles = glob($directory . '/*.php') ?: [];
		foreach ($phpFiles as $phpFile) {
			$fixtures[$phpFile] = true;
			$fileName = basename($phpFile);
			if ($fileName === 'good.php') {
				$hasGood = true;
			} elseif ($fileName === 'bad.php') {
				$hasBad = true;
			} elseif (str_starts_with($fileName, 'custom')) {
				$customCount++;
			}
		}
	}

	$rows[] = [
		'enabled_sniff' => $sniff,
		'has_tests' => count($directories) > 0,
		'fixture_count' => count($fixtures),
		'has_good' => $hasGood,
		'has_bad' => $hasBad,
		'custom_count' => $customCount,
		'directories' => implode(';', array_map(static fn(string $path): string => ltrim(str_replace($root, '', $path), '/'), $directories)),
	];
}

$csv = fopen($csvPath, 'wb');
if ($csv === false) {
	fwrite(STDERR, "Cannot open CSV report for writing: {$csvPath}\n");
	exit(1);
}

fputcsv($csv, ['enabled_sniff', 'has_tests', 'fixture_count', 'has_good', 'has_bad', 'custom_count', 'directories']);
foreach ($rows as $row) {
	fputcsv($csv, [
		$row['enabled_sniff'],
		$row['has_tests'] ? 'true' : 'false',
		(string) $row['fixture_count'],
		$row['has_good'] ? 'true' : 'false',
		$row['has_bad'] ? 'true' : 'false',
		(string) $row['custom_count'],
		$row['directories'],
	]);
}
fclose($csv);

$total = count($rows);
$tested = count(array_filter($rows, static fn(array $row): bool => $row['has_tests'] === true));
$missing = array_values(array_filter($rows, static fn(array $row): bool => $row['has_tests'] === false));
$weak = array_values(array_filter(
	$rows,
	static fn(array $row): bool => $row['has_tests'] === true && (!$row['has_good'] || !$row['has_bad'] || $row['fixture_count'] === 1)
));
$strong = count(array_filter(
	$rows,
	static fn(array $row): bool => $row['has_tests'] === true && $row['has_good'] && $row['has_bad'] && $row['custom_count'] > 0
));

$summary = [];
$summary[] = '# Sniff Coverage Summary';
$summary[] = '';
$summary[] = '- Total enabled sniffs: ' . $total;
$summary[] = '- With dedicated tests: ' . $tested;
$summary[] = '- Missing dedicated tests: ' . count($missing);
$summary[] = '- Weak examples: ' . count($weak);
$summary[] = '- Strong examples (good+bad+custom): ' . $strong;
$summary[] = '';

if ($missing !== []) {
	$summary[] = '## Missing';
	foreach ($missing as $row) {
		$summary[] = '- `' . $row['enabled_sniff'] . '`';
	}
	$summary[] = '';
}

if ($weak !== []) {
	$summary[] = '## Weak Examples';
	foreach ($weak as $row) {
		$summary[] = sprintf(
			'- `%s` (fixtures=%d, good=%s, bad=%s)',
			$row['enabled_sniff'],
			$row['fixture_count'],
			$row['has_good'] ? 'true' : 'false',
			$row['has_bad'] ? 'true' : 'false'
		);
	}
	$summary[] = '';
}

file_put_contents($summaryPath, implode("\n", $summary));

fwrite(STDOUT, "Generated {$csvPath}\n");
fwrite(STDOUT, "Generated {$summaryPath}\n");

File: .dev/sniff-coverage-matrix.py
#!/usr/bin/env python3

from __future__ import annotations

import csv
import re
import subprocess
from collections import defaultdict
from pathlib import Path


ROOT = Path(__file__).resolve().parent.parent
REPORTS = ROOT / ".reports"
SNIFFS_ROOT = ROOT / "tests" / "Sniffs"

SNIFF_RE = re.compile(r"^\s*([A-Z][A-Za-z0-9]*\.[A-Za-z0-9_]+\.[A-Za-z0-9_]+)\s*$")
RULE_REF_RE = re.compile(r"<rule\s+ref=\"([^\"]+)\"")


def run_phpcs_enabled() -> list[str]:
    result = subprocess.run(
        ["vendor/bin/phpcs", "--standard=ruleset.xml", "-e"],
        cwd=ROOT,
        text=True,
        capture_output=True,
        check=True,
    )

    sniffs: list[str] = []
    for line in result.stdout.splitlines():
        match = SNIFF_RE.match(line)
        if match:
            sniffs.append(match.group(1))

    return sorted(set(sniffs))


def build_ref_directory_map() -> dict[str, set[Path]]:
    refs: dict[str, set[Path]] = defaultdict(set)

    for ruleset in SNIFFS_ROOT.glob("**/*.ruleset.xml"):
        content = ruleset.read_text(encoding="utf-8")
        for ref in RULE_REF_RE.findall(content):
            if "." in ref:
                refs[ref].add(ruleset.parent)

    return refs


def summarize_sniff(ref: str, directories: set[Path]) -> dict[str, object]:
    fixture_paths: set[Path] = set()
    has_good = False
    has_bad = False
    custom_count = 0

    for directory in sorted(directories):
        for php_file in directory.glob("*.php"):
            fixture_paths.add(php_file)
            if php_file.name == "good.php":
                has_good = True
            elif php_file.name == "bad.php":
                has_bad = True
            elif php_file.name.startswith("custom"):
                custom_count += 1

    return {
        "enabled_sniff": ref,
        "has_tests": bool(directories),
        "fixture_count": len(fixture_paths),
        "has_good": has_good,
        "has_bad": has_bad,
        "custom_count": custom_count,
        "directories": ";".join(str(path.relative_to(ROOT)) for path in sorted(directories)),
    }


def main() -> None:
    enabled = run_phpcs_enabled()
    refs = build_ref_directory_map()

    REPORTS.mkdir(parents=True, exist_ok=True)
    csv_path = REPORTS / "sniff-coverage-matrix.csv"
    md_path = REPORTS / "sniff-coverage-summary.md"

    rows = [summarize_sniff(ref, refs.get(ref, set())) for ref in enabled]

    with csv_path.open("w", newline="", encoding="utf-8") as handle:
        writer = csv.DictWriter(
            handle,
            fieldnames=[
                "enabled_sniff",
                "has_tests",
                "fixture_count",
                "has_good",
                "has_bad",
                "custom_count",
                "directories",
            ],
        )
        writer.writeheader()
        writer.writerows(rows)

    total = len(rows)
    tested = sum(1 for row in rows if row["has_tests"])
    missing = [row for row in rows if not row["has_tests"]]
    weak = [
        row
        for row in rows
        if row["has_tests"] and (not row["has_good"] or not row["has_bad"] or row["fixture_count"] == 1)
    ]
    strong = [
        row
        for row in rows
        if row["has_tests"] and row["has_good"] and row["has_bad"] and row["custom_count"] > 0
    ]

    with md_path.open("w", encoding="utf-8") as handle:
        handle.write("# Sniff Coverage Summary\n\n")
        handle.write(f"- Total enabled sniffs: {total}\n")
        handle.write(f"- With dedicated tests: {tested}\n")
        handle.write(f"- Missing dedicated tests: {len(missing)}\n")
        handle.write(f"- Weak examples: {len(weak)}\n")
        handle.write(f"- Strong examples (good+bad+custom): {len(strong)}\n\n")

        if missing:
            handle.write("## Missing\n")
            for row in missing:
                handle.write(f"- `{row['enabled_sniff']}`\n")
            handle.write("\n")

        if weak:
            handle.write("## Weak Examples\n")
            for row in weak:
                handle.write(
                    f"- `{row['enabled_sniff']}` (fixtures={row['fixture_count']}, good={row['has_good']}, bad={row['has_bad']})\n"
                )
            handle.write("\n")

    print(f"Generated {csv_path}")
    print(f"Generated {md_path}")


if __name__ == "__main__":
    main()
